std/container/array.d

```diff
@@ -318,7 +318,25 @@ if (!is(Unqual!T == bool))
                 const nbytes = mulu(newLength, T.sizeof, overflow);
                 if (overflow)
                     assert(0);
-                _payload = (cast(T*) realloc(_payload.ptr, nbytes))[0 .. newLength];
+
+                static if (hasIndirections!T)
+                {
+                    auto newPayloadPtr = cast(T*) malloc(nbytes);
+                    newPayloadPtr || assert(false, "std.container.Array.length failed to allocate memory.");
+                    auto newPayload = newPayloadPtr[0 .. newLength];
+                    memcpy(newPayload.ptr, _payload.ptr, _payload.sizeof);
+                    memset(newPayload.ptr + startEmplace, 0,
+                            (newLength - startEmplace) * T.sizeof);
+                    GC.addRange(newPayload.ptr, nbytes);
+                    GC.removeRange(_payload.ptr);
+                    free(_payload.ptr);
+                    _payload = newPayload;
```

Comment:
Just out of curiosity: what's the reason for always allocating a new block, as opposed to, say, just calling GC.addRange / GC.removeRange when the pointer returned by `realloc` changes?  Seems to me that always reallocating might be a little more expensive in the case where `realloc` would have reallocated in-place.

Reply:
I'll have to check when I get time to answer this. I think I largely copied this from a nearby function.